src/kudu/consensus/raft_consensus.cc

```diff
@@ -2226,6 +2226,48 @@ Status RaftConsensus::CheckBulkConfigChangeAndGetNewConfigUnlocked(
           }
           if (IsRaftConfigVoter(server_uuid, committed_config)) {
             num_voters_modified++;
+
+            // If we are in flexi-raft mode, we want to make sure that the number of
+            // voters does not dip below quorum requirements/min-rep-factor
+            // So if we have 6 voters in LEADER region and min-replication-factor/
+            // quorum = Majority(6) = 4, then in healthy state we are expecting 6 voters.
+            // We can safely remove 2 voters and we will still have 4 voters, having enough
+            // for write-availability, but we can't remove another one as that will make #voters=3
+            // which will prevent commit with a MIN-REP-FACTOR=4
+            if (FLAGS_enable_flexi_raft) {
+              const auto& vd_map = committed_config.voter_distribution();
+
+              std::map<std::string, int> voters_in_config_per_region;
+              std::string unused_leader_region;
+              std::string unused_leader_uuid;
+              // Get number of voters in each region
+              GetRegionalCountsFromConfig(
+                  committed_config, unused_leader_uuid, &voters_in_config_per_region,
+                  &unused_leader_region);
+
+              for (const RaftPeerPB& peer : committed_config.peers()) {
+                if (peer.permanent_uuid() != server_uuid) {
+                  continue;
+                }
+
+                // Zeroed in on the peer we are about to remove.
+                const std::string& region = peer.attrs().region();
+                int current_count = voters_in_config_per_region[region];
```

Comment:
Can it so happen that `voters_in_config_per_region` wont have an entry for `region`. New region additions and /or bugs etc? If so, then [ ] will end up default constructing and inserting into `voters_in_config_per_region[region]` with default value of int (0 here) and return it. The subsequent `int future_count = current_count - 1;` will underflow.

Reply:
actually this is a good point. better to be defensive here.

Reply:
this is not possible, because we built GetRegionalCountsFromConfig from the current config and this peer was
part of the committed config (as we have filtered it from the same).